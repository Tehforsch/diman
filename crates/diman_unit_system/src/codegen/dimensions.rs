use crate::dimension_math::BaseDimensions;
use proc_macro2::TokenStream;
use quote::{quote, quote_spanned};

use super::{storage_types::StorageType, Codegen};

impl Codegen {
    pub fn get_dimension_expr(&self, dim: &BaseDimensions) -> TokenStream {
        let dimension_type = &self.defs.dimension_type;
        let field_updates: TokenStream = dim
            .fields()
            .map(|(field, value)| self.get_base_dimension_entry(field, value))
            .collect();
        let span = self.defs.quantity_type.span();
        let none_update = if dim.num_fields() < self.defs.base_dimensions.len() {
            quote! { ..#dimension_type::none() }
        } else {
            quote! {}
        };
        quote_spanned! {span =>
            #dimension_type {
                #field_updates
                #none_update
            }
        }
    }

    pub(crate) fn gen_dimensions(&self) -> TokenStream {
        let dimension_type = &self.defs.dimension_type;
        let quantity_type = &self.defs.quantity_type;
        let defs = self.gen_dimension_definitions();
        quote! {
            pub mod dimensions {
                use super::#dimension_type;
                use super::#quantity_type;
                use super::Exponent;
                #defs
            }
        }
    }

    fn gen_dimension_definitions(&self) -> TokenStream {
        let dimension_type = &self.defs.dimension_type;
        let quantity_type = &self.defs.quantity_type;
        let dimensions: TokenStream = self
            .defs
            .dimensions
            .iter()
            .map(|quantity| {
                let dimension = self.get_dimension_expr(&quantity.dimensions);
                let quantity_type = &self.defs.quantity_type;
                let quantity_name = &quantity.name;
                let span = self.defs.dimension_type.span();
                quote_spanned! {span =>
                    pub type #quantity_name<S> = #quantity_type::<S, { #dimension }>;
                }
            })
            .collect();
        quote! {
            #dimensions
        }
    }

    fn constant_definitions_for_storage_type(&self, type_: &dyn StorageType) -> TokenStream {
        self
            .defs.constants
            .iter()
            .map(|constant| {
                let dimension = self.get_dimension_expr(&constant.dimensions);
                let quantity_type = &self.defs.quantity_type;
                let constant_name = &constant.name;
                let magnitude = constant.magnitude.into_f64();
                let float_type = &type_.base_storage().name;
                let type_ = type_.name();
                // TODO(minor): The allow(clippy::approx_constant)
                // exists to allow definitions of, for example, PI in
                // unit_system calls.  A better solution would
                // probably be to define PI (and possibly some other
                // mathematical constants) for use in the unit_system
                // macro, but this is an easy fix for now.
                quote! {
                    #[allow(clippy::approx_constant)]
                    pub const #constant_name: #quantity_type::<#type_, { #dimension }> = #quantity_type::<#type_, { #dimension }>(#magnitude as #float_type);
                }
            })
            .collect()
    }
}
