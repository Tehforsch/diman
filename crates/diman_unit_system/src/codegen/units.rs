use proc_macro2::TokenStream;
use quote::{format_ident, quote, quote_spanned};
use syn::Type;

use super::storage_types::{FloatType, VectorType};
use crate::types::{Defs, Unit};

impl Defs {
    pub fn gen_unit_constructors(&self) -> TokenStream {
        self.units
            .iter()
            .map(|unit| {
                let dimension = self.get_dimension_expr(&unit.dimensions);
                let vector_impls: TokenStream = self
                    .vector_types()
                    .iter()
                    .map(|vector_type| self.vector_unit_constructor(vector_type, unit, &dimension))
                    .collect();
                let float_impls: TokenStream = self
                    .float_types()
                    .iter()
                    .map(|float_type| self.float_unit_constructor(float_type, unit, &dimension))
                    .collect();
                let conversion_impls: TokenStream = self
                    .storage_types()
                    .map(|storage_type| {
                        let type_ = storage_type.name();
                        let float_type = storage_type.base_storage();
                        self.conversion_impls(type_, float_type, unit, &dimension)
                    })
                    .collect();
                quote! {
                    #conversion_impls
                    #float_impls
                    #vector_impls
                }
            })
            .collect()
    }

    fn conversion_impls(
        &self,
        storage_type: &Type,
        base_type: &FloatType,
        unit: &Unit,
        dimension: &TokenStream,
    ) -> TokenStream {
        let quantity_type = &self.quantity_type;
        let unit_name = &unit.name;
        let conversion_method_name = format_ident!("in_{}", unit_name);
        let magnitude = unit.magnitude;
        let base_type = &base_type.name;
        quote! {
            impl #quantity_type<#storage_type, {#dimension}> {
                pub fn #conversion_method_name(self) -> #storage_type {
                    let magnitude: #base_type = #magnitude as #base_type;
                    self.0 / magnitude
                }
            }
        }
    }

    fn float_unit_constructor(
        &self,
        float_type: &FloatType,
        unit: &Unit,
        quantity_dimension: &TokenStream,
    ) -> TokenStream {
        let Defs { quantity_type, .. } = &self;
        let Unit {
            name: unit_name,
            magnitude,
            ..
        } = unit;
        let name = &float_type.name;
        let span = self.dimension_type.span();
        // Without const_fn_floating_point_arithmetic (https://github.com/rust-lang/rust/issues/57241)
        // we cannot make unit constructors a const fn in general (since it requires the unstable
        // const_fn_floating_point_arithmetic feature). The following allows the constructor with 1.0
        // conversion factor to be const.
        let const_fn = *magnitude == 1.0;
        let fn_def = if const_fn {
            quote! { const fn }
        } else {
            quote! { fn }
        };
        let value = if const_fn {
            quote! { val }
        } else {
            quote! { val * #magnitude as #name }
        };
        quote_spanned! {span =>
            impl #quantity_type<#name, {#quantity_dimension}> {
                pub #fn_def #unit_name(val: #name) -> #quantity_type<#name, {#quantity_dimension}> {
                    #quantity_type::<#name, {#quantity_dimension}>(#value)
                }
            }
        }
    }

    fn vector_unit_constructor(
        &self,
        vector_type: &VectorType,
        unit: &Unit,
        quantity_dimension: &TokenStream,
    ) -> TokenStream {
        let Defs { quantity_type, .. } = &self;
        let Unit {
            name: unit_name,
            magnitude,
            ..
        } = unit;
        let VectorType {
            name,
            float_type,
            num_dims,
            ..
        } = &vector_type;
        let float_type = &float_type.name;
        let fn_args = match num_dims {
            2 => quote! { x: #float_type, y: #float_type },
            3 => quote! { x: #float_type, y: #float_type, z: #float_type },
            _ => unreachable!(),
        };
        let call_args = match num_dims {
            2 => quote! { x, y },
            3 => quote! { x, y, z },
            _ => unreachable!(),
        };
        let span = self.dimension_type.span();
        quote_spanned! {span =>
            impl #quantity_type<#name, {#quantity_dimension}> {
                pub fn #unit_name(#fn_args) -> #quantity_type<#name, {#quantity_dimension}> {
                    #quantity_type::<#name, {#quantity_dimension}>(#name::new(#call_args) * (#magnitude as #float_type))
                }
            }
        }
    }
}
