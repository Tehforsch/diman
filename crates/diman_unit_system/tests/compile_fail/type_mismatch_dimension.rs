#![allow(incomplete_features)]
#![feature(generic_const_exprs, adt_const_params)]
pub mod example_system;
use example_system::units::meters;
use example_system::dimensions::Time;

fn main() {
    let x: Time<f64> = meters.new(1.0);
}
