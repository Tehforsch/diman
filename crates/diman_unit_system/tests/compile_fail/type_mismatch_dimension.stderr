error[E0308]: mismatched types
 --> tests/compile_fail/type_mismatch_dimension.rs:8:24
  |
8 |     let x: Time<f64> = meters.new(1.0);
  |                        ^^^^^^^^^^^^^^^ expected `Dimension { length: 0, time: 1, mass: 0, temperature: 0 }`, found `Dimension { length: 1, time: 0, mass: 0, temperature: 0 }`
  |
  = note: expected constant `Dimension { length: 0, time: 1, mass: 0, temperature: 0 }`
             found constant `Dimension { length: 1, time: 0, mass: 0, temperature: 0 }`
